KAFKA-10847: Set shared outer store to an in-memory store when in-memory stores are supplied

spena opened the pull request:
When users supply in-memory stores for left/outer joins, then the internal shared outer store must be switch to in-memory store too. This will allow users who want to keep all stores in memory to continue doing so.

Tests?
- Added unit tests to validate topology and left/outer joins work fine with an in-memory shared store.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

spena committed:
KAFKA-10847: Set shared outer store to an in-memory store when in-memory stores are supplied

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImplJoin.java:
```diff
@@ -266,17 +267,36 @@ private void assertUniqueStoreNames(final WindowBytesStoreSupplier supplier,
+        } else {
+            builder = Stores.windowStoreBuilder(
+                Stores.inMemoryWindowStore(
+                    storeName + "-memory-store",
```
I think we should use the same name here; the metrics-scope would differentiate rocksdb from in-memory, which is sufficient.

guozhangwang commented on streams/src/test/java/org/apache/kafka/streams/TopologyTest.java:
```diff
@@ -863,6 +920,60 @@ public void streamStreamLeftJoinTopologyWithCustomStoresNames() {
+                "    Processor: KSTREAM-WINDOWED-0000000003 (stores: [in-memory-join-store-other])\n" +
+                "      --> KSTREAM-OUTEROTHER-0000000005\n" +
+                "      <-- KSTREAM-SOURCE-0000000001\n" +
+                "    Processor: KSTREAM-JOINTHIS-0000000004 (stores: [in-memory-join-store-other, KSTREAM-OUTERSHARED-0000000004-memory-store])\n" +
```
Hmm.. this makes me thinking: does it worth "working around" it to move the naming mechanism of the shared store to `sharedOuterJoinWindowStoreBuilder` above such that it always goes along with the other two store's naming patterns? As you can see here, if the store names are not provided but just the store suppliers, the existing stores would use customized name but the shared store would still use system-provided names.

spena replied:
I just moved the name builder inside the `sharedOuterJoinWindowStoreBuilder`. I did not create the name based on the store supplier, though. I would have to append a suffix and I wasn't sure if it would be too long. For instance, if the store supplier is named `in-memory-join-store`, then I would name the shared outer store as `in-memory-join-store-left-shared-join-store`. 

Also, what name should I use for the full outer? The `this` supplier or the `other` supplier? Say suppliers are named `mem-left-store` and `mem-right-store`. Would I end up with `mem-left-store-outer-shared-join-store` or `mem-right-store-outer-shared-join-store`?

What do you think?

guozhangwang replied:
Maybe we can also make it a bit simpler to just rely on the left side: if the supplier is given on the left side (which would always provide the name), then we name the shared as `thisStoreSupplier.name() + "outerJoinSuffix"`; if it is not, then we look into `streamJoinedInternal.storeName()`, and the last resort as `outerJoinStoreGeneratedName`.

guozhangwang commented on streams/src/main/java/org/apache/kafka/streams/kstream/internals/KStreamImplJoin.java:
```diff
@@ -155,7 +156,7 @@ public long get() {
             final String outerJoinStoreName = userProvidedBaseStoreName == null ? outerJoinStoreGeneratedName : userProvidedBaseStoreName + outerJoinSuffix;
 
-            outerJoinWindowStore = Optional.of(sharedOuterJoinWindowStoreBuilder(outerJoinStoreName, windows, streamJoinedInternal));
+            outerJoinWindowStore = Optional.of(sharedOuterJoinWindowStoreBuilder(outerJoinStoreName, windows, streamJoinedInternal, persistent));
```
Just curious: any rationale that we rely on the left store's persistence only to determine the shared store's persistence?

spena replied:
No reasons at all. I just chose one side to decide whether go mem or rocks. Left seems the only side that always is True either is Left or Outer. I was thinking of maybe going mem if in left joins, the left side is mem; or if in outer joins, both sides left/right are mem. What do you think?

guozhangwang replied:
Maybe keeping it simple to always go with left is fine for now, I was just wondering if you have a specific reason for that :)

guozhangwang reviewed (commented):
This is a meta comment: inside the in-memory window store, we organize data as

```
ConcurrentNavigableMap<Long, ConcurrentNavigableMap<Bytes, byte[]>> segmentMap
```

And uses `SortedMap#entrySet()` for `all()` calls, which relies on Byte's LEXICO_COMPARATOR directly. It's worth checking in benchmarks whether this call would be a bottleneck for in-memory stores (for rocksDB, their comparator is native c++ so supposed to be efficient) -- and if it is a bottleneck, we may need to do some quick optimization on the comparator for `all()` specifically.

spena committed:
refactor: move shared store name builder inside the sharedOuterJoinWindowStoreBuilder

guozhangwang reviewed (commented):
@mjsax @spena I think I found the reason why in-memory store perf degrades over time: in `InMemoryWindowStore#put` we have the following

```
            } else if (!retainDuplicates) {
                // Skip if value is null and duplicates are allowed since this delete is a no-op
                segmentMap.computeIfPresent(windowStartTimestamp, (t, kvMap) -> {
                    kvMap.remove(key);
                    if (kvMap.isEmpty()) {
                        segmentMap.remove(windowStartTimestamp);
                    }
                    return kvMap;
                });
            }
```

I.e. as long as `retainDuplicates` is true we would never delete anything even if `value` bytes are null, and hence the map would just keep growing. I think this is just an overlooked bug that we should fix.

spena committed:
Use custom store supplier names for outer shared join names

spena commented:
Thanks @guozhangwang, that seems the culprit of performance problem. In fact, when the performance tests finish, the script takes some time to complete cleaning - I assume it is taking time to clean all consumed memory.

I will work on this fix in another follow-up PR, is that ok? I addressed all comments in this so it is ready to merge if you're ok with it.

guozhangwang commented:
Merged to trunk, thanks @spena
